java-checks/src/main/java/org/sonar/java/symexec/SymbolicEvaluator.java

```diff
@@ -60,308 +63,333 @@
-        if (symbol.isVariableSymbol()) {
-          return new SymbolicValue.SymbolicVariableValue((Symbol.VariableSymbol) symbol);
-        }
+    /**
+     * evaluates a condition.
+     *
+     * a condition is an expression (of type boolean), which can occur in the following places:
+     * - condition of an if statement
+     * - condition of a loop statement: do...while, for, while (but not foreach)
+     * - left operand of a logical and && and or || operators
+     * - first operand of a ternary operator
+     * conditions of switch statement are broader and not evaluated using this method.
+     *
+     * this method evaluates the given condition. if the result is either true or false,
+     * then the execution state is returned.
+     * otherwise, if the result is unknown, then the execution state is split in two branches
+     * with result TRUE and FALSE. whenever possible constraints are also set on the variables
+     * found in the condition.
+     *
+     * @param state execution state
+     * @param tree tree representing the condition
+     * @return output states
+     */
+    private List<ExecutionState> evaluateCondition(ExecutionState executionState, ExpressionTree tree) {
+      List<ExecutionState> result = new ArrayList<>();
+      ExpressionTree expressionTree = removeCastAndParenthesis(tree);
+      if (expressionTree.is(Tree.Kind.CONDITIONAL_AND)) {
+        evaluateConditionConditionalAnd(result, executionState, (BinaryExpressionTree) expressionTree);
+      } else if (expressionTree.is(Tree.Kind.CONDITIONAL_OR)) {
+        evaluateConditionConditionalOr(result, executionState, (BinaryExpressionTree) expressionTree);
+      } else if (expressionTree.is(Tree.Kind.EQUAL_TO)) {
+        evaluateConditionRelationalOperator(result, executionState, (BinaryExpressionTree) expressionTree, SymbolicRelation.EQUAL_TO);
+      } else if (expressionTree.is(Tree.Kind.GREATER_THAN)) {
+        evaluateConditionRelationalOperator(result, executionState, (BinaryExpressionTree) expressionTree, SymbolicRelation.GREATER_THAN);
+      } else if (expressionTree.is(Tree.Kind.GREATER_THAN_OR_EQUAL_TO)) {
+        evaluateConditionRelationalOperator(result, executionState, (BinaryExpressionTree) expressionTree, SymbolicRelation.GREATER_EQUAL);
+      } else if (expressionTree.is(Tree.Kind.LESS_THAN)) {
+        evaluateConditionRelationalOperator(result, executionState, (BinaryExpressionTree) expressionTree, SymbolicRelation.LESS_THAN);
+      } else if (expressionTree.is(Tree.Kind.LESS_THAN_OR_EQUAL_TO)) {
+        evaluateConditionRelationalOperator(result, executionState, (BinaryExpressionTree) expressionTree, SymbolicRelation.LESS_EQUAL);
+      } else if (expressionTree.is(Tree.Kind.NOT_EQUAL_TO)) {
+        evaluateConditionRelationalOperator(result, executionState, (BinaryExpressionTree) expressionTree, SymbolicRelation.NOT_EQUAL);
+      } else if (expressionTree.is(Tree.Kind.IDENTIFIER)) {
```

Comment:
@merciesa if values are dense, then switch on enum will be faster than sequence of if-else-if